src/util/json_expr.cpp

```diff
@@ -19,6 +21,64 @@ Author: Peter Schrammel
 #include "fixedbv.h"
 #include "std_expr.h"
 #include "config.h"
+#include "identifier.h"
+#include "language.h"
+#include "invariant.h"
+
+#include <langapi/mode.h>
+
+static exprt simplify_json_expr(
+  const exprt &src,
+  const namespacet &ns)
+{
+  if(src.id()==ID_constant)
+  {
+    const typet &type=ns.follow(src.type());
+
+    if(type.id()==ID_pointer)
+    {
+      const irep_idt &value=to_constant_expr(src).get_value();
+
+      if(value!=ID_NULL &&
+         (value!=std::string(value.size(), '0') ||
+          !config.ansi_c.NULL_is_zero) &&
+         src.operands().size()==1 &&
+         src.op0().id()!=ID_constant)
+        // try to simplify the constant pointer
+        return simplify_json_expr(src.op0(), ns);
+    }
+  }
+  else if(src.id()==ID_address_of &&
+          src.operands().size()==1 &&
+          src.op0().id()==ID_member &&
+          id2string(to_member_expr(
+            src.op0()).get_component_name()).find("@")!=std::string::npos)
+  {
+    // simplify expressions of the form &member_expr(object, @class_identifier)
+    return simplify_json_expr(src.op0(), ns);
+  }
+  else if(src.id()==ID_address_of &&
+          src.operands().size()==1 &&
+          src.op0().id()==ID_index &&
+          to_index_expr(src.op0()).index().id()==ID_constant &&
+          to_constant_expr(
+            to_index_expr(src.op0()).index()).value_is_zero_string())
+  {
+    // simplify expressions of the form  &array[0]
+    return simplify_json_expr(to_index_expr(src.op0()).array(), ns);
```

Comment:
! `&arr[0]` has type `element*` whereas `arr` has type `element[...]`, might be important?